src/tensor_operations/matrix_decomposition.jl

```diff
@@ -360,46 +360,170 @@ function noinds_error_message(decomp::String)
+  #     i1 = inds(QT)[1]
+  #     i2 = inds(QT)[2]
+  #     r1 = inds(RT)[1]
+  #     newqn = -dir(i2) * flux(i1 => Block(b[1]))
+  #     ITensors.setblockqn!(i2, newqn, b[2])
+  #     ITensors.setblockqn!(r1, newqn, b[2])
+  # end
+  # end
+
+  #
+  #  Undo the combine oepration, to recover all tensor indices.
+  #
+  Q, R = itensor(QT) * dag(CL), itensor(RT) * dag(CR)
+
+  # Conditionally remove dummy indices.
+  if (lre)
+    Q, R = remove_trivial_index(Q, R, vαl, vαr)
+  end
+  #
+  # fix up the tag name for the index between Q and R.
+  #  
   q = commonind(Q, R)
-  settags!(Q, tags, q)
-  settags!(R, tags, q)
-  q = settags(q, tags)
+  settags!(Q, qtag, q)
+  settags!(R, qtag, q)
+  q = settags(q, qtag)
 
-  if isempty(Lis_original)
-    Q *= dag(vLα)
+  return Q, R, q
+end
+
+function rq(A::ITensor, Linds...; kwargs...)
+  qtag::TagSet = get(kwargs, :tags, "Link,rq") #tag for new index between Q and R
+  Lis = commoninds(A, indices(Linds...))
+  Ris = uniqueinds(A, Lis)
+  lre = isempty(Lis) || isempty(Ris)
+  # make a dummy index with dim=1 and incorporate into A so the Lis & Ris can never
+  # be empty.  A essentially becomes 1D after collection.
+  if (lre)
+    A, vαl, vαr, Lis, Ris = add_trivial_index(A, Lis, Ris)
   end
-  if isempty(Ris_original)
-    R *= dag(vRα)
+
+  #
+  #  Use combiners to render A down to a rank 2 tensor ready matrix QR routine.
+  #
+  CL, CR = combiner(Lis...), combiner(Ris...)
+  cL, cR = combinedind(CL), combinedind(CR)
+  AC = A * CR * CL
+  #
+  #  Make sure we don't accidentally pass the transpose into the matrix qr routine.
+  #
+  AC = permute(AC, cL, cR; allow_alias=true)
+  # qr the matrix.
+  RT, QT = NDTensors.rq(tensor(AC); kwargs...)
+
+  #
+  #  Undo the combine oepration, to recover all tensor indices.
+  #
+  R, Q = itensor(RT) * dag(CL), itensor(QT) * dag(CR)
+
+  # Conditionally remove dummy indices.
+  if (lre)
+    R, Q = remove_trivial_index(R, Q, vαl, vαr)
   end
+  #
+  # fix up the tag name for the index between Q and R.
+  #  
+  q = commonind(Q, R)
+  settags!(Q, qtag, q)
+  settags!(R, qtag, q)
+  q = settags(q, qtag)
 
-  return Q, R, q
+  return R, Q, q
+end
+
+# lq is exported by the LinearAlgebra module so we need acknowledge that to avoid
+# intermitent run time errors.
+function lq(A::ITensor, Linds...; kwargs...)
+  Q, L, q = qr(A, uniqueinds(A, Linds...); kwargs...)
+  #
+  # fix up the tag name for the index between Q and R.
+  #  
+  qtag::TagSet = get(kwargs, :tags, "Link,lq") #tag for new index between Q and R
+  settags!(Q, qtag, q)
+  settags!(L, qtag, q)
+  q = settags(q, qtag)
+
+  return L, Q, q
+end
+
+function ql(A::ITensor, Linds...; kwargs...)
+  Q, L, q = rq(A, uniqueinds(A, Linds...); kwargs...)
+  #
+  # fix up the tag name for the index between Q and R.
+  #  
+  qtag::TagSet = get(kwargs, :tags, "Link,ql") #tag for new index between Q and R
+  settags!(Q, qtag, q)
+  settags!(L, qtag, q)
+  q = settags(q, qtag)
+
+  return L, Q, q
```

Comment:
These would have the same kinds of comments/style suggestions as the `qr` function in #1009.

Reply:
All done except the change of function signature. see https://github.com/ITensor/ITensors.jl/pull/1012#issuecomment-1452522151